janitor/utils.py

```diff
@@ -2373,3 +2377,805 @@ def _sub_process_text_result_MultiIndex(index: pd.MultiIndex, result, df):
+                    raise ValueError(
+                        f"""
+                        {new_label} is present in `df` columns.
+                        Kindly provide unique suffixes to create
+                        columns that are not present in `df`.
+                        """
+                    )
+                mapping[common] = new_label
+            left_on = [
+                f"{label}{left_suffix}" if label in common_columns else label
+                for label in left_on
+            ]
+            df = df.rename(columns=mapping)
+        if right_suffix:
+            mapping = {}
+            for common in common_columns:
+                new_label = f"{common}{right_suffix}"
+                if new_label in right.columns:
+                    raise ValueError(
+                        f"""
+                        {new_label} is present in `right` columns.
+                        Kindly provide unique suffixes to create
+                        columns that are not present in `right`.
+                        """
+                    )
+
+                mapping[common] = new_label
+            right_on = [
+                f"{label}{right_suffix}" if label in common_columns else label
+                for label in right_on
+            ]
+            right = right.rename(columns=mapping)
+
+    conditions = [*zip(left_on, right_on, operators)]
+
+    return (
+        df,
+        right,
+        conditions,
+        how,
+        sort_by_appearance,
+    )
+
+
+def _conditional_join_type_check(
+    left_column: pd.Series, right_column: pd.Series
+) -> None:
+
+    """
+    Raise error if column type is not any of
+    numeric, datetime, or string.
+    """
+
+    numeric_type = all(map(is_numeric_dtype, (left_column, right_column)))
+    date_type = all(map(is_datetime64_dtype, (left_column, right_column)))
+    string_type = all(map(is_string_dtype, (left_column, right_column)))
+
+    numeric_date_string = numeric_type, date_type, string_type
+    if any(numeric_date_string):
+        return None, None
+
+    raise ValueError(
+        """
+        conditional_join only supports
+        numeric, date, or string dtypes.
+        """
+    )
+
+
+# code copied from Stack Overflow
+# https://stackoverflow.com/a/47126435/7175713
+def _le_create_ranges(indices, len_right):
+    cum_length = len_right - indices
+    cum_length = cum_length.cumsum()
+    ids = np.ones(cum_length[-1], dtype=int)
+    ids[0] = indices[0]
+    ids[cum_length[:-1]] = indices[1:] - len_right + 1
+    return ids.cumsum()
+
+
+def _ge_create_ranges(indices):
+    cum_length = indices.cumsum()
+    ids = np.ones(cum_length[-1], dtype=int)
+    ids[0] = 0
+    ids[cum_length[:-1]] = -1 * indices[:-1] + 1
+    return ids.cumsum()
+
+
+def _equal_indices(left_c: pd.Series, right_c: pd.Series):
+    """
+    This uses binary search to get indices where
+    `left_c` is exactly equal to `right_c`.
+
+    Returns a tuple of (left_c, right_c)
+    """
```

Comment:
Minor grammatical nitpick.

```suggestion
    """
    Use binary search to get indices where
    `left_c` is exactly equal to `right_c`.

    Returns a tuple of (left_c, right_c)
    """
```